libtransmission/global-ip-cache.h

```diff
@@ -0,0 +1,123 @@
+// This file Copyright © 2023-2023 Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
+
+#include <array>
+#include <atomic>
+#include <condition_variable>
+#include <chrono> // operator ""min, operator ""s, std::chrono::milliseconds
+#include <memory> // std::unique_ptr
+#include <mutex>
+#include <optional>
+#include <shared_mutex>
+#include <string_view> // operator ""sv
+
+#include "net.h"
+#include "timer.h"
+#include "tr-assert.h"
+#include "web.h"
+
+#pragma once
+
+#ifndef __TRANSMISSION__
+#error only libtransmission should #include this header.
+#endif
+
+using namespace std::literals;
+
+struct tr_session;
+
+/**
+ * Cache global IP addresses.
+ *
+ * This class caches 3 useful info:
+ * 1. Whether your machine supports the IP protocol
+ * 2. Source address used for global connections
+ * 3. Global address
+ *
+ * The idea is, if this class successfully cached a source address, that means
+ * you have connectivity to the public internet. And if the global address is
+ * the same with the source address, then you are not behind an NAT.
+ *
+ * Note: This class isn't meant to be accessed by anyone other than tr_session,
+ * so it has no public methods.
+ */
+class tr_global_ip_cache
+{
+public:
+    template<typename T>
+    using array_ip_t = std::array<T, NUM_TR_AF_INET_TYPES>;
+
+    explicit tr_global_ip_cache(tr_session& session_in);
+    ~tr_global_ip_cache();
+    tr_global_ip_cache(tr_global_ip_cache const&) = delete;
+    tr_global_ip_cache(tr_global_ip_cache&&) = delete;
+    tr_global_ip_cache& operator=(tr_global_ip_cache const&) = delete;
+    tr_global_ip_cache& operator=(tr_global_ip_cache&&) = delete;
+
+    bool try_shutdown() noexcept;
+
+    [[nodiscard]] std::optional<tr_address> global_addr(tr_address_type type) const noexcept;
+    [[nodiscard]] std::optional<tr_address> global_source_addr(tr_address_type type) const noexcept;
+
+    [[nodiscard]] tr_address bind_addr(tr_address_type type) const noexcept;
+
+    void update_addr(tr_address_type type) noexcept;
+
+    // Whether this machine supports this IP protocol
+    array_ip_t<std::atomic_bool> has_ip_protocol_ = { true, true };
```

Comment:
Better to make this a constexpr getter so that `has_ip_protocol_` can be private and so that `array_ip_t` doesn't need to be public

```c++
[[nodiscard]] constexpr bool has_ip_protocol(tr_address_type type) const noexcept
{
    return af_data_[type].has_ip_protocol_;
}
```

Reply:
It can't be constexpr since `std::atomic.load()` is not constexpr.

Reply:
adopted in 780d466d1cf91ad62b92e7411bae36a4b5044a93